### PR TITLE
bpo-40737: Fix possible reference leak for sqlite3 initialization

### PR DESCRIPTION
**Pull request by erlend-aasland**
According to the docs, `PyModule_AddObject()` only decrements the
reference count of value _on success_.  Calling code must `Py_DECREF()`
manually on error.

https://bugs.python.org/issue40737

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2020-05-23-00-22-11.bpo-40737.iph-CM.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Handle PyModule_AddObject() errors in sqlite3 module init correctly.
```
Comment (corona10) [inline Misc/NEWS.d/next/Library/2020-05-23-00-22-11.bpo-40737.iph-CM.rst]:
```suggestion
Fix possible reference leak for :mod:`sqlite3` initialization.
```

**Review by corona10 — APPROVED** [review]
lgtm with nit comment.

File: Modules/_sqlite/module.c [hunk 2]
```diff
@@ -346,6 +346,16 @@ static struct PyModuleDef _sqlite3module = {
         NULL
 };
 
+#define ADD_OBJECT(module, name, type)                             \
+do {                                                               \
+    Py_INCREF(&type);                                              \
+    if (PyModule_AddObject(module, name, (PyObject *)&type) < 0) { \
```
Comment (corona10) [inline Modules/_sqlite/module.c]: You can use PyModule_AddType and you don't have to pass for type name for this case :)

File: Modules/_sqlite/module.c [hunk 3]
```diff
@@ -346,6 +346,16 @@ static struct PyModuleDef _sqlite3module = {
         NULL
 };
 
+#define ADD_OBJECT(module, name, type)                             \
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
#define ADD_TYPE(module, type)                             \
```

use PyModule_AddType :)

File: Modules/_sqlite/module.c [hunk 4]
```diff
@@ -366,14 +376,10 @@ PyMODINIT_FUNC PyInit__sqlite3(void)
...
 
-    Py_INCREF(&pysqlite_ConnectionType);
-    PyModule_AddObject(module, "Connection", (PyObject*) &pysqlite_ConnectionType);
-    Py_INCREF(&pysqlite_CursorType);
-    PyModule_AddObject(module, "Cursor", (PyObject*) &pysqlite_CursorType);
-    Py_INCREF(&pysqlite_PrepareProtocolType);
-    PyModule_AddObject(module, "PrepareProtocol", (PyObject*) &pysqlite_PrepareProtocolType);
-    Py_INCREF(&pysqlite_RowType);
-    PyModule_AddObject(module, "Row", (PyObject*) &pysqlite_RowType);
+    ADD_OBJECT(module, "Connection", pysqlite_ConnectionType);
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
    ADD_TYPE(module,  pysqlite_ConnectionType);
```

File: Modules/_sqlite/module.c [hunk 5]
```diff
@@ -366,14 +376,10 @@ PyMODINIT_FUNC PyInit__sqlite3(void)
...
+    ADD_OBJECT(module, "Cursor", pysqlite_CursorType);
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
    ADD_TYPE(module, pysqlite_CursorType);
```

File: Modules/_sqlite/module.c [hunk 6]
```diff
@@ -366,14 +376,10 @@ PyMODINIT_FUNC PyInit__sqlite3(void)
...
+    ADD_OBJECT(module, "PrepareProtocol", pysqlite_PrepareProtocolType);
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
    ADD_TYPE(module, pysqlite_PrepareProtocolType);
```

File: Modules/_sqlite/module.c [hunk 7]
```diff
@@ -366,14 +376,10 @@ PyMODINIT_FUNC PyInit__sqlite3(void)
...
+    ADD_OBJECT(module, "Row", pysqlite_RowType);
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
    ADD_TYPE(module, pysqlite_RowType);
```

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/module.c (hunk 2 above)
Comment (erlend-aasland) [reply Modules/_sqlite/module.c]: Nice, thanks! :) Will fix.

File: Misc/NEWS.d/next/Library/2020-05-23-00-22-11.bpo-40737.iph-CM.rst (hunk 1 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2020-05-23-00-22-11.bpo-40737.iph-CM.rst]: Ah, neat with the .rst syntax. I'll make a note of that for future commits.

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again.

**Comment by erlend-aasland** [comment]
Should probably be backported to 3.9, @corona10.

File: Modules/_sqlite/cursor.h [hunk 8]
```diff
@@ -64,7 +64,7 @@ PyObject* pysqlite_cursor_fetchall(pysqlite_Cursor* self, PyObject* args);
 PyObject* pysqlite_noop(pysqlite_Connection* self, PyObject* args);
 PyObject* pysqlite_cursor_close(pysqlite_Cursor* self, PyObject* args);
 
-int pysqlite_cursor_setup_types(void);
+void pysqlite_cursor_setup_types(void);
```
Comment (corona10) [inline Modules/_sqlite/cursor.h]: https://github.com/python/cpython/blob/f7b1e461567e5e3fa3ba46f589d9edc1b45b2dd0/Objects/typeobject.c#L5454

Since PyTypeReady check wheather the type is ready status.
You don't have to change the init functions.

Comment (corona10) [reply Modules/_sqlite/cursor.h]: Please revert the relevant changes.

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/cursor.h (hunk 8 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.h]: Thanks. The `pysqlite_*_setup_types()` functions could still need a clean-up after this. I guess that would call for a separate issue.

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.h]: Fixed in commit d96c5e40a9286eeaeddedbdd1d6276fb0eb1a142

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

File: Modules/_sqlite/cache.c [hunk 9]
```diff
@@ -263,7 +263,7 @@ static PyMethodDef cache_methods[] = {
 
 PyTypeObject pysqlite_NodeType = {
         PyVarObject_HEAD_INIT(NULL, 0)
-        MODULE_NAME "Node",                             /* tp_name */
+        MODULE_NAME ".Node",                             /* tp_name */
```
Comment (corona10) [inline Modules/_sqlite/cache.c]: This looks like the sperate issue.
Please open the new issue for this and revert the change :)

Thanks alot!

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/cache.c (hunk 9 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cache.c]: Fixed in 21e678871e28d11c3a84a6380b540756f7ff7a7f

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

File: Modules/_sqlite/module.c [hunk 10]
```diff
@@ -346,6 +346,15 @@ static struct PyModuleDef _sqlite3module = {
         NULL
 };
 
+#define ADD_TYPE(module, type)                 \
+do {                                           \
+    Py_INCREF(&type);                          \
```
Comment (corona10) [inline Modules/_sqlite/module.c]:
```suggestion
```

**Review by corona10 — CHANGES_REQUESTED** [review]
Last review :)
It will cause the leak!
https://github.com/python/cpython/blob/a487a39dca4c41305928c7dfdbcb0b3aa344683b/Python/modsupport.c#L693

File: Modules/_sqlite/module.c (hunk 10 above)
Comment (erlend-aasland) [reply Modules/_sqlite/module.c]: Argh, that one snuck back in the revert (commit d96c5e40a9286eeaeddedbdd1d6276fb0eb1a142). Thanks!

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

**Review by corona10 — APPROVED** [review]
lgtm! Thanks for the contribution!

**Comment by erlend-aasland** [comment]
Thanks for you patience and helpful comments, @corona10!
